Added a file that tests *all* elements for too few children.

NSoiffer opened the pull request:
There are 32 tests in the file.

Currently, the test assumes that the structure is preserved. This is still
up in the air (#15). If that changes, then the elements with id="ref_xxx"
need to change, but everything else should remain unchanged.

I have removed the old tests which were more limited (fixes #159)

NSoiffer committed:
Added a file that tests *all* elements for too few children.
There are 32 tests in the file.
Currently, the test assumes that the structure is preserved. This is still
up in the air (#15). If that changes, then the elements with id="ref_xxx"
need to change, but everything else should remain unchanged.
I have removed the old tests which were more limited (fixes #159)

NSoiffer committed:
Turns out mtable and mtr allow zero elements. Fixed tests to reflect that.

fred-wang commented on mathml/presentation-markup/too-few-children.html:
```diff
@@ -0,0 +1,455 @@
+        // )
+      },
+      "features/preconditions for running empty MathML tests"
+    );
```
This should be removed. The test() are really independent from each other, so this will make this one test fail for browsers lacking mathml layout but does not affect at all the result in the other tests.

fred-wang commented on mathml/presentation-markup/too-few-children.html:
```diff
@@ -0,0 +1,455 @@
+      {name:"mtd", n:1},
+      {name:"maction", n:1},
+      {name:"semantics", n:1},
+      {name:"math", n:1}
```
I don't think the tests for n=1 are correct. It seems to me that all these elements are allowed to have no children. We should only keep the elements that have explicit expected number of children in MathML Core.

I would also move mmultiscripts and tabular elements in separate tests.

fred-wang commented on mathml/presentation-markup/too-few-children.html:
```diff
@@ -0,0 +1,455 @@
+            function() {
+              compareLayout(getBox(element.name, i), getBox("ref_" + element.name, i), 0);
+            },
+            "too few children for " + element.name + " (" + i + " " + (i==1 ? "child" : "children") + ")."
```
Again, please use string literal here.

fred-wang commented on mathml/presentation-markup/too-few-children.html:
```diff
@@ -0,0 +1,455 @@
+<head>
+<meta charset="utf-8">
+<title>Test if elements layout properly if they have too few children</title>
+<link rel="help" href="https://mathml-refresh.github.io/mathml-core/#script-and-limit-schemata">
```
We should have links to the section corresponding to each element tested here. And these sections should clearly say how to handle "too few children". AFAIK, the core spec allows mrow-like elements with 0 children and the tabular section is not really well-explained (it defers to HTML).

fred-wang commented on mathml/presentation-markup/too-few-children.html:
```diff
@@ -0,0 +1,455 @@
+        if (i < element.n) {
+          test(
+            function() {
+              compareLayout(getBox(element.name, i), getBox("ref_" + element.name, i), 0);
```
Adding a

    assert_true(MathMLFeatureDetection.has_mspace());
    assert_true(MathMLFeatureDetection.has_mrow());

at the beginning of the function should be enough for now. Note that MathMLFeatureDetection caches the result, so the subsequents calls are fast.

If you want you can also add

     assert_true(MathMLFeatureDetection[`has_${element.name}`]());

so that the detection is stronger when the functions are implemented in MathMLFeatureDetection

fred-wang commented on mathml/presentation-markup/too-few-children.html:
```diff
@@ -0,0 +1,455 @@
+
+  function runTests() {
+
+    // elements being tested
```
I think the description of the array can be improved here. Probably renaming "n" to "minChildCount" would help to understand what these are about.

fred-wang reviewed (changes requested):
overall this looks good to me, but I think some changes should be made first.

Commit message should contain the full URL to https://github.com/mathml-refresh/mathml/issues other github is confused regarding which issues you are referring to.

NSoiffer committed:
Changed epsilon when doing comparision to a small number from 0. Probably
will be needed once code is fixed to match expected behavior according to
the spec.

Also fixed a bug in the script that isn't currently being tickled but
might when more tests get added.
